### DISCUSSION TITLE
Duplicate Spots

### DISCUSSION THREAD

**Discussion by Logiqx** [djong1/gp3s#36] [General]
Some UK spots have duplicates. e.g. I'll list the correct spot names first:

- Southend-on-sea (correct spot)
  - The Ray
  - The Ray Southend on Sea

- Portland Harbour, Weymouth (correct spot)
  - Portland Harbour
  - Weymouth Portland Harbour

Perhaps something can be done to merge them and moving forward can something maybe be done to avoid people creating new spots unnecessarily?

**Comment by djong1** [comment]
done

**Comment by djong1** [comment]
@Logiqx  I have been struggling with this for a while and any advice is welcome. We can merge spots on a manual basis, but this does not scale. When posting a session, I suggest the closest spot based on the coordinates and highest number of sessions a user can override this suggestion.  Automating was not successful due to the following:

- GEO Fence: detect spot that are close to each other and merge -> Problem: you merge spots that have complete different carateristics, for example: green one is on open sea and red one is a lake and spots are within 1km distance
<img width="195" alt="image" src="https://user-images.githubusercontent.com/25177991/183633793-3461c096-7119-4624-ab83-d328c6e75518.png">

- Naming: to many variants to detect duplicates. 
- quit locations and purely use centered coordinates (disruptive change for the site and it's concept).

**Comment by Logiqx** [comment]
This reminds me of fuzzy matching processes that I have automated in the past but where there is the possibility of "false positives".

I'm having a bit of a busy day so not had the time to give this the thought it deserves but I feel there must be a solution, even if it requires a bit of upfront effort. Maybe it can crowd-source the intial effort within community?

One solution might be to semi-automate:

- Automatically produce list of possible duplicates via geofencing (or simple proximity checks)
- Get help from the community; either a general request or via targetted emails (e.g. people who post for the spots in question), asking for clarification
- Maintain a list of spots that have been confirmed as definitely not the same, so they can be removed from output of future automated checks
- Run the automated check periodically but only list spots that contain sessions since the last run.

Another idea to consider might be to send a follow-up e-mail to someone after they create a potential duplicate. Could simply ask for confirmation that the spot is not a duplicate of the existing one.

A third idea which is pretty easy is to create a KML for each country with pin markers for all of the spots. Spots that are within say 1km of each other (and not confirmed as different) could be have red pins and other spots >1 km apart given blue pins. This would allow people to quickly glance at their own country to see if there are potential duplicates. Combined with my idea of maintaining a list of spots that are definitely not the same, it would be possible to colour the "good" spots blue so that in theory, red spots are the ones that need confirmation one way or the other.

A fourth thought is that I think certain spots might be more prone to this issue than others;
- Portland Harbour vs Weymouth
- Southend-on-Sea vs The Ray

The website could potentially refuse to create a new spot so close to the existing one.

Building on the fourth idea (kinda a fifth idea) would be to only allow people who have already posted to the existing spot to create a new variant in close proximity. They can still be aked if it's a duplicate but they are more likely to realise their mistake imho. This will prevent non-regulars from creating a new spot because they went to sail "The Ray" for the first time and did not realise the established spot is "Southend-on-Sea". The same happens in Weymouth... Weymouth, Portland and Portland Harbour are essentially synonymous so a first-time poster is prone to creating a new spot.

**Comment by Logiqx** [comment]
One further idea.

If a duplicate spot has been created and a session posted to it but other people posted to the existing spot on the same day, could send an e-mail to the entire group saying that a new spot has been created and asking for confirmation as to whether it is a duplicate spot.

Could simply re-direct them to some kind of form capture to collate the responses.

Hopefully the number of times that this happens wouldn;t make it too onorous to keep on top of it?

**Comment by jhdebruin** [comment]
In general i think a lot of the duplicated spots is something we have collected in the 18 years the platform has been running.... to improve it first of all we have to somehow use an automated procedure to determine the possible cases, this could be done at geographic positions or maybe a bit easier also by comparing spot names.  Based on this issue list we could target the riders from the specific spots and ask them what spot should be merged....

For new spots we have also think of a solution, the same approach , so based on geographical positions but also on naming convention....

**Comment by Logiqx** [comment]
@jhdebruin agree there are two steps, firstly fixing existing issues and secondly avoiding future issues. It's likely that the approach could be similar for both.

@djong1 can you produce a bespoke CSV extract of existing spots? I'd be happy to have a play and try some ideas.

e.g. spot id, spot name, country id, location (lon + lat), creation date

I don't know if you have already calculated spot locations from existing sessions or if location data is only stored for the sessions.

I also don't know if you store the creation date for each spot. IDs are sequential and already provide some insight about chronology.

**Comment by djong1** [comment]
@Logiqx done, see your mail. thanks

**Comment by Logiqx** [comment]
Ok, I've created some reports to help us to understand the extent of the issue.

It only works for spots with lat + lon coordinates but it determines "clusters" where spots are within 1km of each other.

Here is an example for Mauritius:

![image](https://user-images.githubusercontent.com/17449078/184335932-dfb99314-cd27-4df0-9ee7-5e5c4fb7a2c8.png)

Clicking on the KML links will show maps of the clusters - yellow pins show the GP3S spot coordinates:

![image](https://user-images.githubusercontent.com/17449078/184336116-5368cad9-3551-4032-8aff-f0c308aa98ad.png)

The Netherlands has the most clusters, let's just look at the first two:

![image](https://user-images.githubusercontent.com/17449078/184336827-ef633845-503b-4cca-8861-56330d3d8e4d.png)

The first cluster is obviously the same spot:

![image](https://user-images.githubusercontent.com/17449078/184336488-a87f6099-8b86-41ef-8eba-12012ff16e8d.png)

The second can be considered to be different spots:

![image](https://user-images.githubusercontent.com/17449078/184336672-f3c9287f-b458-48f9-821f-6efac84be118.png)

I'll write my thoughts about how it can be tackled in a second comment. Here's the link to the reports and KML files:

https://logiqx.github.io/gp3s-query/spot-clusters/

p.s. You'll need Google Earth installed to use the KML links. Just click the link and open the KL file that is downloaded.

**Comment by Logiqx** [comment]
There are only 153 spot clusters. Some clusters are clearly the same spot, others are clearly different spots.

I don't see a simple + safe way to automatically make decisions about individual clusters, without a disproportionate amount of development effort and testing. Having looked at multiple clusters, I can see that name matching would be fraught with difficulty. This style of geo-matching provides visibility of clusters with minimal development effort.

I think a manual review of these 153 clusters, followed by an automated / scripted update makes sense to fix existing issues.

Things to consider when doing a scripted update:

- What about peoples spots list?
- What about home spots?
- What about achievements, spot records, etc?
- May need some clever rebuilds of spot rankings as well?

The above points aside, I think the majority of the clusters can be manually reviewed in quite a short period of time.

Two lists can quickly be produced during this review process:

1) Spot IDs to merge, including the desired spot name for the merged results.
2) List of spot IDs that are distinct, despite being in close proximity.

The spots to merge can subsquently be scripted; ensuring personal spots, achievements, spot rankings are updated accordingly.

I suggest standardising names to things like "Portland Harbour, Weymouth" or "The Ray, Southend-on-Sea" when merging spots with similar (or differing) names. Should probably use the original spot ID (i.e. lowest value) for the merged spots.

The spot IDs to be kept separate can be recorded in a "verified" list. These can subsequently be ignored by this style of reporting.

Avoiding future duplicates is a slightly different activity (imho) because it's a case of determining whether the new spot is a duplicate. I have a variety of thoughts on that topic but they can wait for now.

Finding the existing issues benefits from looking at "clusters" of spots and is slightly more complex which is what this initial investigation helps with, I think.

**Comment by Logiqx** [comment]
@djong1 + @jhdebruin  please have a browse of the reports and let me know your thoughts.

https://logiqx.github.io/gp3s-query/spot-clusters/

**Comment by Logiqx** [comment]
Some of these duplicates are really long standing issues. One example is one of my favorite holiday spots which I've noticed multiple times in the past:

![image](https://user-images.githubusercontent.com/17449078/184341851-7902266c-d5a9-4b05-84d9-ab4e333b1c5a.png)

The spot IDs are 672 and 2545 so the duplication has existing for a long time.

In this instance the standardised name would be "Prasonisi, Rhodos".

Likewise for Vasiliki and Lefkada which could be a merged spot called "Vasiliki, Lefkada":

![image](https://user-images.githubusercontent.com/17449078/184342624-d52fb880-fd94-41bc-b352-21833b6bc27c.png)

On the north side of Rhodes there are two spots which can be called ; "Trianda, Rhodos" and "Ixia Bay, Rhodos":

![image](https://user-images.githubusercontent.com/17449078/184342972-f3fa6175-c2bb-4420-90e7-c15ceaa98f75.png)

I believe a well scripted batch update (ensuring things like user spots, home spot, achievements and spot rankings are updated accordingly) would be beneficial for many spots around the world.

**Comment by jhdebruin** [comment]
@Logiqx Thnx, this extra info will give some fuel for the discussions, will read it all and share the idea's here ,

**Comment by Logiqx** [comment]
It makes for quite interesting browsing.

Here's what appears to be a cluster of 5 spots which are potentially 2 spots:

![image](https://user-images.githubusercontent.com/17449078/184344413-e477556e-ce74-4976-8ddd-593049c38abe.png)

**Comment by Logiqx** [comment]
Just giving this a bit of a poke. Perhaps something can be done this year to address the spot duplicates?

I think the tool I wrote to generate KML files should make it relatively easy to work through the historical dupes.
